### ISSUE TITLE
`this` scope not working in async functions in Babel 6

### ISSUE THREAD

**Issue by cb1kenobi** [babel/babel#2686]
Node.js 5.0.0
babel-cli: 6.0.12
babel-plugin-transform-async-to-generator: 6.0.2
babel-preset-es2015: 6.0.12

```javascript
class Foo {
    constructor() {
        this.name = 'baz';
    }

    bar() {
		console.log('hi from bar');
        return async (...args) => {
			try {
				console.log('hi from async fn');
				console.log(this);
				console.log('name =', this.name);
			} catch (e) {
				console.error(e);
			}
            return this.name;
        };
    }
}

let foo = new Foo();
foo.bar()()
	.then((name) => {
	    if (name === 'baz') {
	        console.log('worked');
	    } else {
	        console.log('expected baz, got', name);
	    }
	});
```

```bash
$ npm install babel-cli babel-preset-es2015 babel-plugin-transform-async-to-generator
$ node node_modules/babel-cli/bin/babel-node --presets es2015 --plugins transform-async-to-generator test.js
hi from bar
hi from async fn
undefined
[TypeError: Cannot read property 'name' of undefined]
```

We can see in the generated code that the returned function is not bound to the correct `this` scope:

```javascript
// <snip>
_createClass(Foo, [{
	key: 'bar',
	value: function bar() {
		console.log('hi from bar');
		return function () {
			var _this = this; // <---------- this should probably move up a scope or this function needs to be bind()'d

			return _asyncToGenerator(_regenerator2.default.mark(function _callee() {
				return _regenerator2.default.wrap(function _callee$(_context) {
					while (1) switch (_context.prev = _context.next) {
						case 0:
							try {
								console.log('hi from async fn');
								console.log(_this);
								console.log('name =', _this.name);
							} catch (e) {
								console.error(e);
							}
							return _context.abrupt('return', _this.name);

						case 2:
						case 'end':
							return _context.stop();
					}
				}, _callee, _this);
			}))();
		};
	}
}]);
// <snip>
```

Side question: why is this using regenerator? Shouldn't transform-async-to-generator be using remap-async-to-generator and thus using an actual generator?

**Comment by sebmck** [comment]
> Side question: why is this using regenerator? Shouldn't transform-async-to-generator be using remap-async-to-generator and thus using an actual generator?

Because you're using the `es2015` preset which enables regenerator for normal generators.

**Comment by sebmck** [comment]
Should be fixed via #2725 and #2708

**Comment by cb1kenobi** [comment]
Confirmed, it works! Thanks @sebmck!
